{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1ad6f071-6cde-41c9-b4b6-e6c51420b17d",
   "metadata": {},
   "source": [
    "# Notebook to edit metadata for Qiita: Cancer dataset from Liao\n",
    "Guide: https://qiita.ucsd.edu/static/doc/html/checklist-for-ebi-ena-submission.html#checklist"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a64d7106-41b3-4355-9f2e-2920d0744420",
   "metadata": {},
   "source": [
    "# Questions:\n",
    "- Do we want to add in from here: https://www.nature.com/articles/s41597-021-00860-8 or should we just link the paper?\n",
    "    - tblbc.csv: Daily measurements of white blood cells, platelets and red blood cells for 1,278 patients\n",
    "    - tbltemperature.csv: temperatures for 1,249 patients\n",
    "    - tblInfectionsCidPapers.csv: The day of positive blood cultures for 426 patients and microbes (genera Enterococcus, Escherichia, Klebsiella, Enterobacter, Pseudomonas, Stenotrophomonas, and Citrobacter) analysed in previous publications from our team\n",
    "    - tbldrug.csv: Timing and route of drug administration for 1,278 patients\n",
    "    - tblhctmeta.csv: The day and source of HCT for 1,278 patients\n",
    "    \n",
    "- If we follow the sample info exactly, (https://qiita.ucsd.edu/static/doc/html/checklist-for-ebi-ena-submission.html#checklist) we are missing the following: \n",
    "    - elevation, lat, long\n",
    "    - physical_specimen_location\n",
    "    - collection_timestamp\n",
    "        - We have a de-identifyed time stamp, so similar\n",
    "        \n",
    "- If we follow the prep info exactly, (https://qiita.ucsd.edu/static/doc/html/checklist-for-ebi-ena-submission.html#checklist) we are missing the following:\n",
    "    - barcode\n",
    "        - Only info I could find in the paper: A unique 12-base Golay barcode (Ns) precedes the primers for sample identification, and one to eight additional nucleotides were placed in front of the barcode to offset the sequencing of the primers. \n",
    "    - Two sets of primers: primer designed to amplify the V4-V5: 563 F (5′-nnnnnnnn-NNNNNNNNNNNN-AYTGGGYDTAAAGNG-3′) and 926 R (5′-nnnnnnnn-NNNNNNNNNNNN-CCGTCAATTYHTTTRAGT-3′). \n",
    "        - Do these need to be run as two different studies? I've never had a study with two primers before\n",
    "    - Experimental_design_description\n",
    "    - Center name \n",
    "    - center_project_name\n",
    "    - library_construction_protocol\n",
    "- Under Patient_ID some are labled 'pt_with_samples_ID'\n",
    "    - I'm not sure what this means so I assumed this was still a sample, but wasn't sure if these were controls?\n",
    "- Was there any particular reason some of the Patient_IDs were labled with 'FMT' to start, should these be pooled?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "90989e22-4fac-4dda-bcc6-20334c14d74d",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import random"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1d6c586-232f-4b0b-ab77-c5d3a1b563d6",
   "metadata": {},
   "source": [
    "### Import file paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "10dfc62a-03a0-4044-8933-2fba66072fa3",
   "metadata": {},
   "outputs": [],
   "source": [
    "fn_path = '22_07_13_QiitaCancer_Liao_filenames.txt'\n",
    "meta_path = 'tblASVsamples_06222022.csv'"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6333a85b-5a99-4d27-aa91-df0fe2822efc",
   "metadata": {},
   "source": [
    "### Convert into list/df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "bda41358-3b8b-4d66-9b70-8d3d3746e64a",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['1000A..pool483_R1.fastq.gz', '1000A..pool483_R2.fastq.gz', '1000B..pool483_R1.fastq.gz']\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Consistency</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>DayRelativeToNearestHCT</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000A</td>\n",
       "      <td>1000</td>\n",
       "      <td>0</td>\n",
       "      <td>formed</td>\n",
       "      <td>SRR11414397</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000B</td>\n",
       "      <td>1000</td>\n",
       "      <td>5</td>\n",
       "      <td>liquid</td>\n",
       "      <td>SRR11414992</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000C</td>\n",
       "      <td>1000</td>\n",
       "      <td>15</td>\n",
       "      <td>liquid</td>\n",
       "      <td>SRR11414991</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID PatientID  Timepoint Consistency    Accession   BioProject  \\\n",
       "0    1000A      1000          0      formed  SRR11414397  PRJNA545312   \n",
       "1    1000B      1000          5      liquid  SRR11414992  PRJNA545312   \n",
       "2    1000C      1000         15      liquid  SRR11414991  PRJNA545312   \n",
       "\n",
       "   DayRelativeToNearestHCT AccessionShotgun Pool  \n",
       "0                     -9.0              NaN  483  \n",
       "1                     -4.0              NaN  483  \n",
       "2                      6.0              NaN  483  "
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Filename list\n",
    "fn = []\n",
    "with open(fn_path) as file:\n",
    "    while (line := file.readline().rstrip()):\n",
    "        fn.append(line)\n",
    "print(fn[:3])\n",
    "\n",
    "#Metadata df\n",
    "meta = pd.read_csv(meta_path)\n",
    "meta[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e230e77-d649-4d63-81ab-351229fa201a",
   "metadata": {},
   "source": [
    "## Sample Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "0b70fc60-7135-4038-a483-de8002a7878c",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>sample_name</th>\n",
       "      <th>host_subject_id</th>\n",
       "      <th>Timepoint</th>\n",
       "      <th>Consistency</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>DayRelativeToNearestHCT</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "      <th>taxon_id</th>\n",
       "      <th>scientific_name</th>\n",
       "      <th>env_biome</th>\n",
       "      <th>env_feature</th>\n",
       "      <th>env_material</th>\n",
       "      <th>env_package</th>\n",
       "      <th>empo_1</th>\n",
       "      <th>empo_2</th>\n",
       "      <th>empo_3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000A</td>\n",
       "      <td>1000</td>\n",
       "      <td>0</td>\n",
       "      <td>formed</td>\n",
       "      <td>SRR11414397</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>-9.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>256318</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>human-associated habitat</td>\n",
       "      <td>feces</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal-associated</td>\n",
       "      <td>Animal distal gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000B</td>\n",
       "      <td>1000</td>\n",
       "      <td>5</td>\n",
       "      <td>liquid</td>\n",
       "      <td>SRR11414992</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>-4.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>256318</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>human-associated habitat</td>\n",
       "      <td>feces</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal-associated</td>\n",
       "      <td>Animal distal gut</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000C</td>\n",
       "      <td>1000</td>\n",
       "      <td>15</td>\n",
       "      <td>liquid</td>\n",
       "      <td>SRR11414991</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>6.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>256318</td>\n",
       "      <td>metagenome</td>\n",
       "      <td>urban biome</td>\n",
       "      <td>human-associated habitat</td>\n",
       "      <td>feces</td>\n",
       "      <td>human-gut</td>\n",
       "      <td>Host-associated</td>\n",
       "      <td>Animal-associated</td>\n",
       "      <td>Animal distal gut</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  sample_name host_subject_id  Timepoint Consistency    Accession  \\\n",
       "0       1000A            1000          0      formed  SRR11414397   \n",
       "1       1000B            1000          5      liquid  SRR11414992   \n",
       "2       1000C            1000         15      liquid  SRR11414991   \n",
       "\n",
       "    BioProject  DayRelativeToNearestHCT AccessionShotgun Pool  taxon_id  \\\n",
       "0  PRJNA545312                     -9.0              NaN  483    256318   \n",
       "1  PRJNA545312                     -4.0              NaN  483    256318   \n",
       "2  PRJNA545312                      6.0              NaN  483    256318   \n",
       "\n",
       "  scientific_name    env_biome               env_feature env_material  \\\n",
       "0      metagenome  urban biome  human-associated habitat        feces   \n",
       "1      metagenome  urban biome  human-associated habitat        feces   \n",
       "2      metagenome  urban biome  human-associated habitat        feces   \n",
       "\n",
       "  env_package           empo_1             empo_2             empo_3  \n",
       "0   human-gut  Host-associated  Animal-associated  Animal distal gut  \n",
       "1   human-gut  Host-associated  Animal-associated  Animal distal gut  \n",
       "2   human-gut  Host-associated  Animal-associated  Animal distal gut  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s = meta.copy()\n",
    "\n",
    "#sample_name\n",
    "s.rename(columns = {'SampleID':'sample_name'}, inplace = True)\n",
    "\n",
    "#host_subject_id\n",
    "s.rename(columns = {'PatientID':'host_subject_id'}, inplace = True)\n",
    "\n",
    "#taxon_id and scientific_name\n",
    "s[\"taxon_id\"] = 256318\n",
    "s[\"scientific_name\"] = 'metagenome'\n",
    "\n",
    "#env\n",
    "s[\"env_biome\"] = 'urban biome'\n",
    "s[\"env_feature\"] = 'human-associated habitat'\n",
    "s[\"env_material\"] = 'feces'\n",
    "s[\"env_package\"] = 'human-gut'\n",
    "\n",
    "#empo\n",
    "s[\"empo_1\"] = 'Host-associated'\n",
    "s[\"empo_2\"] = 'Animal-associated'\n",
    "s[\"empo_3\"] = 'Animal distal gut'\n",
    "\n",
    "s[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "db2bebe5-bf91-4fba-881e-633a0a2e1095",
   "metadata": {},
   "source": [
    "## Prep Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "d277adb5-397b-41c7-b8c4-9b6ce2c405ab",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "12535\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "      <th>run_prefix</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>platform</th>\n",
       "      <th>target_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000A</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414397</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000A..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000B</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414992</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000B..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000C</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414991</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000C..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID PatientID    Accession   BioProject AccessionShotgun Pool  \\\n",
       "0    1000A      1000  SRR11414397  PRJNA545312              NaN  483   \n",
       "1    1000B      1000  SRR11414992  PRJNA545312              NaN  483   \n",
       "2    1000C      1000  SRR11414991  PRJNA545312              NaN  483   \n",
       "\n",
       "       run_prefix instrument_model  platform target_gene  \n",
       "0  1000A..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "1  1000B..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "2  1000C..pool483   Illumina MiSeq  Illumina    16S rRNA  "
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "p = meta.copy()\n",
    "\n",
    "#sample_name\n",
    "s.rename(columns = {'SampleID':'sample_name'}, inplace = True)\n",
    "\n",
    "#run_prefix\n",
    "p[\"run_prefix\"] = p[\"SampleID\"] + str('..pool') + p[\"Pool\"]\n",
    "\n",
    "#instrument_model\n",
    "p[\"instrument_model\"] = 'Illumina MiSeq'\n",
    "\n",
    "#platform\n",
    "p[\"platform\"] = 'Illumina'\n",
    "\n",
    "#target_gene\n",
    "p['target_gene'] = '16S rRNA'\n",
    "\n",
    "p.drop(['Timepoint', 'Consistency', 'DayRelativeToNearestHCT'], inplace=True, axis=1)\n",
    "print(len(p))\n",
    "p[:3]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7e004e28-9d79-4434-b69c-ec397966a80f",
   "metadata": {},
   "source": [
    "### Split Prep Information"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ad3ea8c2-4c3f-49e5-99d5-14b0dc125deb",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "3408\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "      <th>run_prefix</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>platform</th>\n",
       "      <th>target_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1004A</td>\n",
       "      <td>FMT.0002</td>\n",
       "      <td>SRR11414975</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1004A..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1004B</td>\n",
       "      <td>FMT.0002</td>\n",
       "      <td>SRR11414972</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1004B..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1083A</td>\n",
       "      <td>FMT.0023</td>\n",
       "      <td>SRR11414929</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>550</td>\n",
       "      <td>1083A..pool550</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID PatientID    Accession   BioProject AccessionShotgun Pool  \\\n",
       "0    1004A  FMT.0002  SRR11414975  PRJNA545312              NaN  483   \n",
       "1    1004B  FMT.0002  SRR11414972  PRJNA545312              NaN  483   \n",
       "2    1083A  FMT.0023  SRR11414929  PRJNA545312              NaN  550   \n",
       "\n",
       "       run_prefix instrument_model  platform target_gene  \n",
       "0  1004A..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "1  1004B..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "2  1083A..pool550   Illumina MiSeq  Illumina    16S rRNA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "735        68\n",
      "621        59\n",
      "698        58\n",
      "761        58\n",
      "730        57\n",
      "           ..\n",
      "516.BMT     1\n",
      "500         1\n",
      "762         1\n",
      "553         1\n",
      "668.fmt     1\n",
      "Name: Pool, Length: 151, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "#Split FMT. cases out\n",
    "fmt = p[p.apply(lambda x: 'FMT' in x.PatientID, axis=1)]\n",
    "fmt = fmt.reset_index(drop=True)\n",
    "\n",
    "print(len(fmt))\n",
    "display(fmt[:3])\n",
    "print(fmt.loc[:]['Pool'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "b28fa122-c0aa-4b60-9aff-bd1912503af7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of groups: 6 of ~ 500 samples\n"
     ]
    }
   ],
   "source": [
    "#Randomly split remaning cases keeping pool numbers aligned\n",
    "\n",
    "#Change this variable depending on the size per group\n",
    "aprx_size_lists = 500\n",
    "\n",
    "pool_df = fmt.loc[:]['Pool'].value_counts() \n",
    "all_pools = list(pool_df.index)\n",
    "pool_groups = []\n",
    "\n",
    "while len(all_pools) > int(aprx_size_lists/50):\n",
    "    current_group = []\n",
    "    sum_current_group = 0\n",
    "    while sum_current_group < aprx_size_lists and len(all_pools)>=1:\n",
    "        selected_pool = random.sample(all_pools, 1)\n",
    "        count_in_pool = pool_df[selected_pool][0]\n",
    "        sum_current_group += count_in_pool\n",
    "        current_group.append(selected_pool[0])\n",
    "        all_pools.remove(selected_pool[0])\n",
    "    pool_groups.append(current_group)\n",
    "    #print(sum_current_group)\n",
    "print('Total number of groups:', len(pool_groups), 'of ~', aprx_size_lists, 'samples')\n",
    "fmt_pool_groups = pool_groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "eb47b3e9-4821-4c69-9004-0af69f9e2fed",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "186\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "      <th>run_prefix</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>platform</th>\n",
       "      <th>target_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1001</td>\n",
       "      <td>pt_with_samples_1001_1002_1003_1004_1005_1006_...</td>\n",
       "      <td>SRR11414988</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>535</td>\n",
       "      <td>1001..pool535</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1002</td>\n",
       "      <td>pt_with_samples_1001_1002_1003_1004_1005_1006_...</td>\n",
       "      <td>SRR11414981</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>535</td>\n",
       "      <td>1002..pool535</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1005</td>\n",
       "      <td>pt_with_samples_1001_1002_1003_1004_1005_1006_...</td>\n",
       "      <td>SRR11414971</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>203</td>\n",
       "      <td>1005..pool203</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID                                          PatientID    Accession  \\\n",
       "0     1001  pt_with_samples_1001_1002_1003_1004_1005_1006_...  SRR11414988   \n",
       "1     1002  pt_with_samples_1001_1002_1003_1004_1005_1006_...  SRR11414981   \n",
       "2     1005  pt_with_samples_1001_1002_1003_1004_1005_1006_...  SRR11414971   \n",
       "\n",
       "    BioProject AccessionShotgun Pool     run_prefix instrument_model  \\\n",
       "0  PRJNA545312              NaN  535  1001..pool535   Illumina MiSeq   \n",
       "1  PRJNA545312              NaN  535  1002..pool535   Illumina MiSeq   \n",
       "2  PRJNA545312              NaN  203  1005..pool203   Illumina MiSeq   \n",
       "\n",
       "   platform target_gene  \n",
       "0  Illumina    16S rRNA  \n",
       "1  Illumina    16S rRNA  \n",
       "2  Illumina    16S rRNA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Split pt cases out\n",
    "pt = p[p.apply(lambda x: 'pt' in x.PatientID, axis=1)]\n",
    "pt = pt.reset_index(drop=True)\n",
    "print(len(pt))\n",
    "display(pt[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "9ff09929-ddb1-4d2f-9a2c-16110a326599",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "8941\n",
      "12535\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SampleID</th>\n",
       "      <th>PatientID</th>\n",
       "      <th>Accession</th>\n",
       "      <th>BioProject</th>\n",
       "      <th>AccessionShotgun</th>\n",
       "      <th>Pool</th>\n",
       "      <th>run_prefix</th>\n",
       "      <th>instrument_model</th>\n",
       "      <th>platform</th>\n",
       "      <th>target_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1000A</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414397</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000A..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>1000B</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414992</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000B..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1000C</td>\n",
       "      <td>1000</td>\n",
       "      <td>SRR11414991</td>\n",
       "      <td>PRJNA545312</td>\n",
       "      <td>NaN</td>\n",
       "      <td>483</td>\n",
       "      <td>1000C..pool483</td>\n",
       "      <td>Illumina MiSeq</td>\n",
       "      <td>Illumina</td>\n",
       "      <td>16S rRNA</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  SampleID PatientID    Accession   BioProject AccessionShotgun Pool  \\\n",
       "0    1000A      1000  SRR11414397  PRJNA545312              NaN  483   \n",
       "1    1000B      1000  SRR11414992  PRJNA545312              NaN  483   \n",
       "2    1000C      1000  SRR11414991  PRJNA545312              NaN  483   \n",
       "\n",
       "       run_prefix instrument_model  platform target_gene  \n",
       "0  1000A..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "1  1000B..pool483   Illumina MiSeq  Illumina    16S rRNA  \n",
       "2  1000C..pool483   Illumina MiSeq  Illumina    16S rRNA  "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Remaining cases (w/o FMT and pt)\n",
    "p0 = p[p.apply(lambda x: 'pt' not in x.PatientID, axis=1)]\n",
    "p0 = p0[p0.apply(lambda x: 'FMT' not in x.PatientID, axis=1)]\n",
    "p0 = p0.reset_index(drop=True)\n",
    "print(len(p0))\n",
    "print(8941 + 186 + 3408)\n",
    "display(p0[:3])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "3bed626f-8ea9-4d34-94a7-a66634045eb4",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Total number of groups: 17 of ~ 500 samples\n"
     ]
    }
   ],
   "source": [
    "#Randomly split remaning cases keeping pool numbers aligned\n",
    "\n",
    "#Change this variable depending on the size per group\n",
    "aprx_size_lists = 500\n",
    "\n",
    "pool_df = p0.loc[:]['Pool'].value_counts() \n",
    "all_pools = list(pool_df.index)\n",
    "pool_groups = []\n",
    "\n",
    "while len(all_pools) > int(aprx_size_lists/50):\n",
    "    current_group = []\n",
    "    sum_current_group = 0\n",
    "    while sum_current_group < aprx_size_lists and len(all_pools)>=1:\n",
    "        selected_pool = random.sample(all_pools, 1)\n",
    "        count_in_pool = pool_df[selected_pool][0]\n",
    "        sum_current_group += count_in_pool\n",
    "        current_group.append(selected_pool[0])\n",
    "        all_pools.remove(selected_pool[0])\n",
    "    pool_groups.append(current_group)\n",
    "    #print(sum_current_group)\n",
    "print('Total number of groups:', len(pool_groups), 'of ~', aprx_size_lists, 'samples')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "563745af-17fe-4c46-a6d8-a1b0af7ed452",
   "metadata": {},
   "source": [
    "## Export df as csv"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "ae48232d-036d-427a-b9ea-bd46972cdf9e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Sample\n",
    "s.to_csv('Liao_sample_info.tsv',  sep='\\t', index = False)\n",
    "\n",
    "#Prep\n",
    "#FMT\n",
    "for group in range(0, len(fmt_pool_groups)):\n",
    "    px = fmt[fmt['Pool'].isin(fmt_pool_groups[group])]\n",
    "    name = 'prep/Liao_prep_fmt_' + str(group+1) + '.tsv'\n",
    "    px.to_csv(name,  sep='\\t', index = False)\n",
    "\n",
    "#pt\n",
    "pt.to_csv('prep/Liao_prep_pt.tsv',  sep='\\t', index = False)\n",
    "\n",
    "#Remaining Cases\n",
    "for group in range(0, len(pool_groups)):\n",
    "    px = p0[p0['Pool'].isin(pool_groups[group])]\n",
    "    name = 'prep/Liao_prep_' + str(group+1) + '.tsv'\n",
    "    px.to_csv(name,  sep='\\t', index = False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "db031b3f-49cd-4220-98ef-67d863ebcc13",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
